File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.js
var shortestPathBinaryMatrix = function(grid) {
    let n = grid.length
    let directions = [[-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1]]
    let queue = [[0, 0]]
    if (grid[0][0] !== 0 || grid[n - 1][n - 1] !== 0) return -1;
    let path = 1

    function isValid(i, j) {
        if (i < 0 || j < 0 || i == n || j == n) return false
        return true
    }

    while (queue.length) {
        let size = queue.length

        for (let k = 0; k < size; k++) {
            let [i, j] = queue.shift()
            if (i == n - 1 && j == n - 1) return path

            for (let [x, y] of directions) {
                let new_x = x + i
                let new_y = y + j

                if (isValid(new_x, new_y) && grid[new_x][new_y] == 0) {
                    grid[new_x][new_y] = 1
                    queue.push([new_x, new_y])
                }
            }
        }
        path += 1
    }

    return -1
};

File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        directions = [[-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1]]
        path = 1
        queue = [(0, 0)]
        n = len(grid)

        if (grid[0][0] != 0):
            return -1


        def isValid(i, j):
            if (i < 0 or j < 0 or i == n or j == n or grid[i][j]):
                return False
            return True

        while (queue):
            size = len(queue)

            for _ in range(size):
                i, j = queue.pop(0)
                if (i == n - 1 and j == n - 1):
                    return path

                for x, y in directions:
                    row = i + x
                    col = j + y

                    if (isValid(row, col)):
                        grid[row][col] = 1
                        queue.append((row, col))
            path += 1

        return -1



        